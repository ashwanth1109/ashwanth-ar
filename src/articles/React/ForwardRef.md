## Forward Ref

Ref forwarding is a technique for automatically passing a ref through a component to one of its children. It is an opt-in feature that lets some components take a ref they receive, and pass it further down (in other words, “forward” it) to a child.

```js
const FancyButton = React.forwardRef((props, ref) => (
  <button ref={ref} className="FancyButton">
    {props.children}
  </button>
));

// You can now get a ref directly to the DOM button:
const ref = React.createRef();
<FancyButton ref={ref}>Click me!</FancyButton>;
```

The second ref argument only exists when you define a component with React.forwardRef call. Regular function or class components don’t receive the ref argument, and ref is not available in props either.

<hr/>

### Refs in HOC

Refs will not get passed through a HOC along with props. That’s because ref is not a prop. Like key, it’s handled differently by React. If you add a ref to a HOC, the ref will refer to the outermost container component, not the wrapped component.

```js
function logProps(Component) {
  class LogProps extends React.Component {
    componentDidUpdate(prevProps) {
      console.log("old props:", prevProps);
      console.log("new props:", this.props);
    }

    render() {
      const { forwardedRef, ...rest } = this.props;

      // Assign the custom prop "forwardedRef" as a ref
      return <Component ref={forwardedRef} {...rest} />;
    }
  }

  // Note the second param "ref" provided by React.forwardRef.
  // We can pass it along to LogProps as a regular prop, e.g. "forwardedRef"
  // And it can then be attached to the Component.
  return React.forwardRef((props, ref) => {
    return <LogProps {...props} forwardedRef={ref} />;
  });
}
```

We can explicitly forward refs to the inner FancyButton component using the React.forwardRef API. React.forwardRef accepts a render function that receives props and ref parameters and returns a React node.

<hr/>

### SOURCE CODE - forwardRef.js

```js
import { REACT_FORWARD_REF_TYPE, REACT_MEMO_TYPE } from "shared/ReactSymbols";

import warningWithoutStack from "shared/warningWithoutStack";

export default function forwardRef<Props, ElementType: React$ElementType>(
  render: (props: Props, ref: React$Ref<ElementType>) => React$Node
) {
  if (__DEV__) {
    if (render != null && render.$$typeof === REACT_MEMO_TYPE) {
      warningWithoutStack(
        false,
        "forwardRef requires a render function but received a `memo` " +
          "component. Instead of forwardRef(memo(...)), use " +
          "memo(forwardRef(...))."
      );
    } else if (typeof render !== "function") {
      warningWithoutStack(
        false,
        "forwardRef requires a render function but was given %s.",
        render === null ? "null" : typeof render
      );
    } else {
      warningWithoutStack(
        // Do not warn for 0 arguments because it could be due to usage of the 'arguments' object
        render.length === 0 || render.length === 2,
        "forwardRef render functions accept exactly two parameters: props and ref. %s",
        render.length === 1
          ? "Did you forget to use the ref parameter?"
          : "Any additional parameter will be undefined."
      );
    }

    if (render != null) {
      warningWithoutStack(
        render.defaultProps == null && render.propTypes == null,
        "forwardRef render functions do not support propTypes or defaultProps. " +
          "Did you accidentally pass a React component?"
      );
    }
  }

  return {
    $$typeof: REACT_FORWARD_REF_TYPE,
    render
  };
}
```

<hr/>
<br/>
<br/>

```js
if (render != null && render.$$typeof === REACT_MEMO_TYPE) {
  warningWithoutStack(
    false,
    "forwardRef requires a render function but received a `memo` " +
      "component. Instead of forwardRef(memo(...)), use " +
      "memo(forwardRef(...))."
  );
}
```

Check if there is something to render and if its a memoized component, then throw out a warning message. So, basically pass in a forwardedRef component to be memoized and not the other way around.

<hr/>

```js
if (typeof render !== "function") {
  warningWithoutStack(
    false,
    "forwardRef requires a render function but was given %s.",
    render === null ? "null" : typeof render
  );
}
```

Check for render function.

<hr/>

```js
warningWithoutStack(
  // Do not warn for 0 arguments because it could be due to usage of the 'arguments' object
  render.length === 0 || render.length === 2,
  "forwardRef render functions accept exactly two parameters: props and ref. %s",
  render.length === 1
    ? "Did you forget to use the ref parameter?"
    : "Any additional parameter will be undefined."
);
```

Check if render function has 0 or 2 parameters. If not, then throw appropriate warnings.

<hr/>

```js
warningWithoutStack(
  render.defaultProps == null && render.propTypes == null,
  "forwardRef render functions do not support propTypes or defaultProps. " +
    "Did you accidentally pass a React component?"
);
```

Does this mean that when you use forwardRef, you cannot use propTypes or defaultProps? That doesnt make sense?!!
