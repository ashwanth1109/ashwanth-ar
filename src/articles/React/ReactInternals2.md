# A Deep Dive into React Internals - Part 2

## createRef

```jsx
// an immutable object with a single mutable value
export function createRef(): RefObject {
  const refObject = {
    current: null
  };
  if (__DEV__) {
    Object.seal(refObject);
  }
  return refObject;
}
```

`__DEV__` => global variable to track whether development or production

`Object.seal()` => This method seals an object preventing new properties from being added to it and marking all existing properties as non-configurable. Values of present properties can still be changed as long as they are writable.

```jsx
export type RefObject = {|
  current: any
|};
```

## ReactCreateEventComponent

```js
import type {
  ReactEventComponent,
  ReactEventResponder
} from "shared/ReactTypes";
import { enableEventAPI } from "shared/ReactFeatureFlags";

import { REACT_EVENT_COMPONENT_TYPE } from "shared/ReactSymbols";
```

### What are polyfills? (from MDN docs)

A polyfill is a piece of code (usually JavaScript on the Web) used to provide modern functionality on older browsers that do not natively support it.

The reason for why polyfills are not used exclusively is for better functionality and better performance. Native implementations of APIs can do more and are faster than polyfills.

```js
let hasBadMapPolyfill;
```

```js
if (__DEV__) {
  hasBadMapPolyfill = false;
  try {
    const frozenObject = Object.freeze({});
    const testMap = new Map([[frozenObject, null]]);
    const testSet = new Set([frozenObject]);
    // This is necessary for Rollup to not consider these unused.
    // https://github.com/rollup/rollup/issues/1771
    // TODO: we can remove these if Rollup fixes the bug.
    testMap.set(0, 0);
    testSet.add(0);
  } catch (e) {
    // TODO: Consider warning about bad polyfills
    hasBadMapPolyfill = true;
  }
}
```

Rollup is a module bundler for JavaScript which compiles small pieces of code into something larger and more complex, such as a library or application.

```js
export function createEventComponent(
  responder: ReactEventResponder,
  displayName: string
): ?ReactEventComponent {
  if (enableEventAPI) {
    // We use responder as a Map key later on. When we have a bad
    // polyfill, then we can't use it as a key as the polyfill tries
    // to add a property to the object.
    if (__DEV__ && !hasBadMapPolyfill) {
      Object.freeze(responder);
    }
    const eventComponent = {
      $$typeof: REACT_EVENT_COMPONENT_TYPE,
      displayName: displayName,
      props: null,
      responder: responder
    };
    if (__DEV__) {
      Object.freeze(eventComponent);
    }
    return eventComponent;
  }
}
```

## ReactChildren

```js
import invariant from "shared/invariant";
import warning from "shared/warning";
import {
  getIteratorFn,
  REACT_ELEMENT_TYPE,
  REACT_PORTAL_TYPE
} from "shared/ReactSymbols";

import { isValidElement, cloneAndReplaceKey } from "./ReactElement";
import ReactDebugCurrentFrame from "./ReactDebugCurrentFrame";
```

What are these separators for?

```js
const SEPARATOR = ".";
const SUBSEPARATOR = ":";
```

Also, what key are they trying to escape here?

Probably similar to the key we provide when mapping through an array to generate html elements for each item in the array.

```js
/**
 * Escape and wrap key so it is safe to use as a reactid
 *
 * @param {string} key to be escaped.
 * @return {string} the escaped key.
 */
function escape(key) {
  const escapeRegex = /[=:]/g;
  const escaperLookup = {
    "=": "=0",
    ":": "=2"
  };
  const escapedString = ("" + key).replace(escapeRegex, function(match) {
    return escaperLookup[match];
  });

  return "$" + escapedString;
}
```

### To be continued
