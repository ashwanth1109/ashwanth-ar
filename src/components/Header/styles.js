// @flow

import styled, { css } from "styled-components";
import { ifProp } from "styled-tools";

import { Logo as StyledLogo } from "pages/styles";
import { colors, fonts } from "styles/constants";

export const Header = styled.div`
  width: 100%;
  position: fixed;
  background: ${colors.darkNight};
  z-index: 1;
`;

export const VisibleHeader = styled.div`
  position: relative;
  z-index: 1;
  background: ${colors.darkNight};
  height: 76px;
  display: flex;
  justify-content: flex-end;
  align-items: flex-end;
  padding: 16px 24px;
`;

export const Logo = styled(StyledLogo)`
  z-index: 2;
  width: 60px;
  height: 60px;
  margin-top: 8px;
  position: absolute;
  left: 16px;
  cursor: pointer;
`;

export const Menu = styled.div`
  display: flex;
  height: 40px;
`;

export const MenuOption = styled.div`
  padding: 0 3vw;
  display: flex;
  color: ${colors.pureWhite};
  font-family: ${fonts.primary};
  font-size: 20px;
  border-bottom: 5px solid
    ${ifProp("selected", colors.goldenYellow, colors.slabGray)};
  align-items: center;
  height: 100%;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
  background-color: ${ifProp("selected", colors.slabGray, "")};
  cursor: pointer;
  transition: 0.5s ease-in-out;
  :hover {
    ${ifProp(
      "selected",
      null,
      css`
        border-color: ${colors.pureWhite};
      `
    )}
  }
`;

export const Contact = styled.div`
  height: ${ifProp("show", "60px", "0px")};
  width: 100vw;
  transition: 0.5s ease-in-out height;
  position: relative;
`;

export const Tray = styled.div`
  background-color: ${colors.slabGray};
  position: absolute;
  width: 100vw;
  height: 60px;
  top: ${ifProp("show", "0px", "-60px")};
  transition: 0.5s ease-in-out top;
  font-size: ${ifProp("smallText", "20px", "32px")};
  font-family: ${fonts.primary};
  display: flex;
  align-items: center;
`;

export const IconContainer = styled.div`
  margin-left: 16px;
  display: flex;
  align-items: center;
  cursor: pointer;
`;

export const Icon = styled.img`
  height: 40px;
  margin-right: 16px;
`;

export const IconTitle = styled.div`
  text-decoration: underline;
`;

export const HamburgerMenu = styled.img`
  height: 30px;
  cursor: pointer;
  position: absolute;
  top: 24px;
  right: 16px;
  z-index: 2;
`;

export const CollapsibleMenu = styled.div`
  position: absolute;
  z-index: 0;
  top: 76px;
  right: 0px;
  width: 200px;
  height: 300px;
  background: ${colors.darkNight};
  border: 2px solid ${colors.pureWhite};
`;

export const HamburgerMenuOption = styled(MenuOption)`
  width: 100%;
  height: 60px;
  border-bottom: 0px;
  color: ${ifProp("selected", colors.goldenYellow, colors.pureWhite)};
  :hover {
    background-color: ${colors.slabGray};
  }
`;
