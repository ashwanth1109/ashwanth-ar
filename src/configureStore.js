// @flow

import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import { connectRoutes } from "redux-first-router";

import page from "./pageReducer";

const routesMap = {
  HOME: "/",
  ABOUT: "/about",
  TECH: "/tech",
  MUSIC: "/music",
  BOOKS: "/books"
};

export default function configureStore(preloadedState) {
  const { reducer, middleware, enhancer } = connectRoutes(routesMap);

  const rootReducer = combineReducers({ page, location: reducer });
  const middlewares = applyMiddleware(middleware);

  // For devtools extension
  const composeEnhancers =
    typeof window === "object" && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
      : compose;

  const enhancers = composeEnhancers(enhancer, middlewares);

  const store = createStore(rootReducer, preloadedState, enhancers);

  return { store };
}
