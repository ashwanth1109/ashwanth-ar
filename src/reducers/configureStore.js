// @flow

import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import { connectRoutes } from "redux-first-router";

import page from "./pageReducer";

import app from "reducers";

const routesMap = {
  HOME: "/",
  ABOUT: "/about",
  TECH: "/tech",
  UNIFIZE: "/tech/unifize-chat",
  TECH_ARTICLES: "/tech/:article",
  MUSIC: "/music",
  MUSIC_ARTICLES: "/music/:article",
  CREDO: "/credo"
};

export default function configureStore(preloadedState) {
  const { reducer, middleware, enhancer } = connectRoutes(routesMap);

  const rootReducer = combineReducers({ page, location: reducer, app });
  const middlewares = applyMiddleware(middleware);

  // For devtools extension
  const composeEnhancers =
    typeof window === "object" && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
      : compose;

  const enhancers = composeEnhancers(enhancer, middlewares);

  const store = createStore(rootReducer, preloadedState, enhancers);

  return { store };
}
