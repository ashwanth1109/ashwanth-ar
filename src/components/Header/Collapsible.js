// @flow

import React from "react";
import { connect } from "react-redux";
import Link from "redux-first-router-link";

import { CollapsibleMenu, HamburgerMenuOption } from "./styles";
import { headerOptions } from "data";
import { string } from "utils";
import { toggleCollapsibleMenu } from "actions/ui";

type Props = {
  currentPage: any,
  showMenu: boolean,
  toggleMenu: Function
};

const Collapsible = ({ currentPage, showMenu, toggleMenu }: Props) => {
  return (
    <>
      <CollapsibleMenu showMenu={showMenu}>
        {headerOptions.map((option, id) => (
          <Link
            key={id}
            to={{ type: option.id, payload: {} }}
            onClick={toggleMenu}
          >
            <HamburgerMenuOption selected={option.id === currentPage}>
              {string.capitalizeString(option.id)}
            </HamburgerMenuOption>
          </Link>
        ))}
      </CollapsibleMenu>
    </>
  );
};

const mapStateToProps = ({ app }) => ({
  showMenu: app.ui.collapsibleMenu
});

export default connect(
  mapStateToProps,
  { toggleMenu: toggleCollapsibleMenu }
)(Collapsible);
