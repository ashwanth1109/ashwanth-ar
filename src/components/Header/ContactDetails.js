// @flow

import React from "react";
import { connect } from "react-redux";

import { Contact, Tray, IconContainer, IconTitle, Icon } from "./styles";

import { iconPack0 } from "assets/icons";
import { Container } from "styles";

type Props = {
  show: boolean,
  width: number
};

const ContactDetails = ({ show, width }: Props) => {
  return (
    <Contact show={show}>
      <Tray show={show} smallText={width < 600}>
        <Container justify={width > 600 ? "space-around" : "center"}>
          Connect with me:
          {iconPack0.map((icon, id) => (
            <IconContainer
              key={id}
              onClick={() => window.open(icon.link, "_blank")}
            >
              <Icon src={icon.src} />
              {width > 600 && <IconTitle>{icon.name}</IconTitle>}
            </IconContainer>
          ))}
        </Container>
      </Tray>
    </Contact>
  );
};

const mapStateToProps = ({ app }) => ({
  width: app.viewport.w
});

export default connect(mapStateToProps)(ContactDetails);
