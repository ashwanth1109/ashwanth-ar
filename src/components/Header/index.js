// @flow

import React, { useState } from "react";
import Link from "redux-first-router-link";
import { connect } from "react-redux";

import {
  Header as StyledHeader,
  VisibleHeader,
  Logo,
  Menu,
  MenuOption,
  HamburgerMenu
} from "./styles";

import logo from "assets/logo.svg";
import menu from "assets/menu.svg";
import { toggleCollapsibleMenu } from "actions/ui";

import { headerOptions } from "data";
import { string } from "utils";
import ContactDetails from "./ContactDetails";
import Collapsible from "./Collapsible";

type Props = {
  currentPage: string,
  width: Number,
  toggleShowMenu: Function
};

const Header = ({ currentPage, width, toggleShowMenu }: Props) => {
  const [contactMenu, setContactMenu] = useState(false);
  console.log(width);

  return (
    <>
      <StyledHeader>
        <ContactDetails show={contactMenu} />
        <Logo src={logo} onClick={() => setContactMenu(prev => !prev)} />
        <VisibleHeader>
          {width > 500 && (
            <Menu>
              {headerOptions.map((option, id) => (
                <Link
                  key={id}
                  to={{
                    type: option.id,
                    payload: {}
                  }}
                >
                  <MenuOption selected={option.id === currentPage}>
                    {string.capitalizeString(option.id)}
                  </MenuOption>
                </Link>
              ))}
            </Menu>
          )}
          <HamburgerMenu src={menu} onClick={toggleShowMenu} />
        </VisibleHeader>
        {width <= 500 && <Collapsible currentPage={currentPage} />}
      </StyledHeader>
    </>
  );
};

const mapStateToProps = ({ location, app }) => ({
  currentPage: location.type,
  width: app.viewport.w
});

export default connect(
  mapStateToProps,
  { toggleShowMenu: toggleCollapsibleMenu }
)(Header);
