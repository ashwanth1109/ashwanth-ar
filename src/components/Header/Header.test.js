// @flow

import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

import Header from "./";
import Collapsible from "./Collapsible";
import ContactDetails from "./ContactDetails";

import configureStore from "reducers/configureStore";

const { act } = renderer;
const { store } = configureStore();

test("Header renders correctly", () => {
  let component;

  act(() => {
    component = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>
    );
  });

  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test("Collapsible renders correctly", () => {
  let component;

  act(() => {
    component = renderer.create(
      <Provider store={store}>
        <Collapsible currentPage="TEST" />
      </Provider>
    );
  });

  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test("ContactDetails renders correctly", () => {
  let component;

  act(() => {
    component = renderer.create(
      <Provider store={store}>
        <ContactDetails show={true} />
      </Provider>
    );
  });

  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
