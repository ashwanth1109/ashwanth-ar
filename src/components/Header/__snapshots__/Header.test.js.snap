// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible renders correctly 1`] = `
<div
  className="sc-jAaTju iXRHdH"
>
  <a
    href="/"
    onClick={[Function]}
  >
    <div
      className="sc-iAyFgw sc-jDwBTQ kZbTRC"
      selected={false}
    >
      Home
    </div>
  </a>
  <a
    href="/about"
    onClick={[Function]}
  >
    <div
      className="sc-iAyFgw sc-jDwBTQ kZbTRC"
      selected={false}
    >
      About
    </div>
  </a>
  <a
    href="/tech"
    onClick={[Function]}
  >
    <div
      className="sc-iAyFgw sc-jDwBTQ kZbTRC"
      selected={false}
    >
      Tech
    </div>
  </a>
  <a
    href="/music"
    onClick={[Function]}
  >
    <div
      className="sc-iAyFgw sc-jDwBTQ kZbTRC"
      selected={false}
    >
      Music
    </div>
  </a>
  <a
    href="/credo"
    onClick={[Function]}
  >
    <div
      className="sc-iAyFgw sc-jDwBTQ kZbTRC"
      selected={false}
    >
      Credo
    </div>
  </a>
</div>
`;

exports[`ContactDetails renders correctly 1`] = `
<div
  className="sc-hSdWYo fufjhz"
>
  <div
    className="sc-eHgmQL bJSUZN"
  >
    <div
      className="sc-bwzfXH loAYuZ"
    >
      Connect with me:
      <div
        className="sc-cvbbAY iXcTdC"
        onClick={[Function]}
      >
        <img
          className="sc-jWBwVP NHLKx"
          src="github.svg"
        />
      </div>
      <div
        className="sc-cvbbAY iXcTdC"
        onClick={[Function]}
      >
        <img
          className="sc-jWBwVP NHLKx"
          src="linkedin.svg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Header renders correctly 1`] = `
<div
  className="sc-eNQAEJ eTFuzd"
>
  <div
    className="sc-hSdWYo dVIlZw"
  >
    <div
      className="sc-eHgmQL dTDcYH"
    >
      <div
        className="sc-bwzfXH loAYuZ"
      >
        Connect with me:
        <div
          className="sc-cvbbAY iXcTdC"
          onClick={[Function]}
        >
          <img
            className="sc-jWBwVP NHLKx"
            src="github.svg"
          />
        </div>
        <div
          className="sc-cvbbAY iXcTdC"
          onClick={[Function]}
        >
          <img
            className="sc-jWBwVP NHLKx"
            src="linkedin.svg"
          />
        </div>
      </div>
    </div>
  </div>
  <img
    className="sc-ckVGcZ sc-kEYyzF eKxdWv"
    onClick={[Function]}
    src="logo.svg"
  />
  <div
    className="sc-hMqMXs ggocWj"
  />
</div>
`;
